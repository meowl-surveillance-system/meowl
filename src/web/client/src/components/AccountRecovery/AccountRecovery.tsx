import React, {
  Component,
  FormEvent,
  ChangeEvent,
  ComponentState,
} from "react";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Container from "@material-ui/core/Container";

interface Props {}
interface State {
  username: string;
  successMessage: string;
  failMessage: string;
}

export default class AccountRecovery extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      username: "",
      successMessage: "",
      failMessage: "",
    };
  }

  onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    try {
      event.preventDefault();
      const requestOptions = {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          username: this.state.username,
        }),
      };
      const res = await fetch("/auth/beginPasswordReset", requestOptions);
      const msg = await res.json();
      if (msg === "Successfully sent password reset email") {
        this.setState({ successMessage: msg, failMessage: "" });
      } else {
        this.setState({ failMessage: msg, successMessage: "" });
      }
    } catch (e) {
      console.error(e);
    }
  };

  onChange = async (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    this.setState({
      [event.target.name]: event.target.value,
    } as ComponentState);
  };

  render() {
    return (
      <Container component="main" maxWidth="xs">
        <Grid
          container
          spacing={2}
          direction="column"
          alignItems="center"
          justify="center"
          style={{ minHeight: "100vh" }}
        >
          {this.state.failMessage && (
            <Typography style={{ color: "red" }}>
              {this.state.failMessage}
            </Typography>
          )}
          {this.state.successMessage && (
            <Typography style={{ color: "green" }}>
              {this.state.successMessage}
            </Typography>
          )}
          <Typography component="h1" variant="h5">
            Account Recovery
          </Typography>
          <Typography>
            {
              "Forgot your account's password? Well damn. Enter your username and we'll send you a recovery link to the associated e-mail"
            }
          </Typography>
          <form noValidate onSubmit={this.onSubmit}>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="username"
              label="Username"
              onChange={this.onChange}
            />
            <Button fullWidth type="submit" variant="contained" color="primary">
              Send Recovery Email
            </Button>
          </form>
        </Grid>
      </Container>
    );
  }
}
