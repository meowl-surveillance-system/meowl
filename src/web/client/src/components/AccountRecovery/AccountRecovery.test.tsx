import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import AccountRecovery from "./AccountRecovery";

describe("AccountRecovery component", () => {
  let AccountRecoveryWrapper: ShallowWrapper<{}, {}, AccountRecovery>;
  let eventMock: any;
  let preventDefaultMock: jest.Mock;
  let mockSuccessResponse: string;
  let mockJsonPromise: Promise<string>;
  let mockFetchPromise: Promise<object>;
  beforeEach(() => {
    mockSuccessResponse = "Successfully sent password reset email";
    mockJsonPromise = Promise.resolve(mockSuccessResponse);
    mockFetchPromise = Promise.resolve({ json: () => mockJsonPromise });
    jest
      .spyOn(window, "fetch")
      .mockImplementation(() => mockFetchPromise as Promise<Response>);
    preventDefaultMock = jest.fn();
    eventMock = {
      preventDefault: preventDefaultMock,
      target: {
        name: "username",
        value: "beeetch",
      },
    };
    AccountRecoveryWrapper = shallow(<AccountRecovery />);
  });
  test("renders AccountRecovery component", () => {
    const { container } = render(
      <Router>
        <AccountRecovery />
      </Router>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should set successMessage on successful resolve", async () => {
    await AccountRecoveryWrapper.instance().onSubmit(eventMock);
    expect(AccountRecoveryWrapper.instance().state.successMessage).toEqual(
      mockSuccessResponse,
    );
  });
  it("should not set failMessage on promise reject", async () => {
    await AccountRecoveryWrapper.instance().onSubmit(eventMock);
    expect(AccountRecoveryWrapper.instance().state.failMessage).toEqual("");
  });
  it("should set password field when onChange is called", () => {
    AccountRecoveryWrapper.instance().onChange(eventMock);
    expect(AccountRecoveryWrapper.instance().state.username).toEqual(
      eventMock.target.value,
    );
  });
});
