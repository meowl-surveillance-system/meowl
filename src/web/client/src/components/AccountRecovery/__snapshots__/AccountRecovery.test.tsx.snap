// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountRecovery component renders AccountRecovery component 1`] = `
<main
  class="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    style="min-height: 100vh;"
  >
    
    
    <h1
      class="MuiTypography-root MuiTypography-h5"
    >
      Account Recovery
    </h1>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Forgot your account's password? Well damn. Enter your username and we'll send you a recovery link to the associated e-mail
    </p>
    <form
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          Username
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="username"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-200 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-202"
            >
              <span>
                Username
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Send Recovery Email
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</main>
`;
